File: 2016/day16/main.hs
import Control.Monad (forM_)

find :: (a -> Bool) -> [a] -> a
find p (x:xs) | p x       = x
              | otherwise = find p xs

generate :: Int -> String -> String
generate l = take l . find ((>= l) . length) . iterate expand where
  expand s = s ++ '0' : map tr (reverse s)
  tr '0' = '1'
  tr '1' = '0'

checksum :: String -> String
checksum = find (odd . length) . iterate shrink where
  shrink (x:y:xs) | x == y = '1' : shrink xs
                  | x /= y = '0' : shrink xs
  shrink []                = []

input :: String
input = "10001110011110000"

main = forM_ [("One", 272), ("Two", 35651584)] $ \(part, l) ->
  putStrLn $ (("Part " ++ part ++ ": ") ++) $ checksum $ generate l input

File: 2016/day16/main.rb

input = "10001110011110000"
#len = 272
len = 35651584

#input = "10000"
#len = 20

def expand(string)
  string + "0" + string.reverse.tr('01','10')
end

while input.size < len
  input = expand(input)
  #puts "expanded to: #{input}"
end

input = input[0, len]

while input.size.even?
  input.gsub!(/../) do |cs|
    cs[0] == cs[1] ? '1' : '0'
  end
  #puts "smalled to: #{input}"
end

puts input
